pkg/bgpv1/gobgp/test.go

```diff
@@ -0,0 +1,27 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package gobgp
+
+import (
+	"context"
+	"fmt"
+	"net/netip"
+	"testing"
+
+	gobgp "github.com/osrg/gobgp/v3/api"
+)
+
+// GetExistingPeerForTest is a test only method to fetch the underlying gobgp.Peer.
+func (g *GoBGPServer) GetExistingPeerForTest(peerAddress string, peerASN int64) (*gobgp.Peer, error) {
+	if !testing.Testing() {
```

Comment:
This is new in go 1.21: https://pkg.go.dev/testing#Testing which means this method can be used for testing but can't be called normally.

For testing this it means it can test the actual password, which seemed easier than fitting into the testing in `pkg/bgpv1/test`, but I'm not hugely attached to this, happy to either keep the API only test (i.e. not test the actual password in unit tests) or potentially something else.

Reply:
Woah, hold on. Importing `"testing"` has the side effect of declaring all testing CLI flags, which we definitely don't want on the main Cilium agent binary importing this package. This is why `testutils` packages exist. Better would be to implement `GetExistingPeer()` as a free function in a separate testutils subpackage that takes a `GoBGPServer` as an argument.

Looks to me like testing.Testing() was introduced to prevent users from pulling in testutils from non-test code, not the other way around.

Reply:
Oh, not quite sure what I was thinking. A `testutils` package doesn't work without making something public which probably shouldn't be. It would be possible to do with build tags, but that makes development annoying (`go test` on its own doesn't work).

I've just reverted this bit, it's possible to infer based on whether the password is set or not if things worked. It would be better to test the rest in an integration test, but it doesn't seem there's one for the BGP control plane (yet?).

Reply:
> A `testutils` package doesn't work without making something public which probably shouldn't be.

That's fine. In your proposal, GoBGPServer and GetExistingPeerForTest were both exported, so not sure what the difference would be. Which symbols are you worried about specifically? Cilium as a project is not concerned about exporting symbols in general -- the only package with an `internal` dir is hive, all other packages are importable by other modules. You could opt for an `internal/testutils` subpackage if this concerns you.

> It would be possible to do with build tags

After spending weeks removing integration and privileged build tags from tests in favor of envs, that's a hard no from me. :sweat_smile: We want to be able to build all code with a single `go test`. arch-specific files aside.

Reply:
> That's fine. In your proposal, GoBGPServer and GetExistingPeerForTest were both exported, so not sure what the difference would be. Which symbols are you worried about specifically? Cilium as a project is not concerned about exporting symbols in general -- the only package with an internal dir is hive, all other packages are importable by other modules. You could opt for an internal/testutils subpackage if this concerns you.

Something in `pkg/bgpv1/gobgp` would have to be exported so something in a testutils can return it, that's why I was trying to add `GetExistingPeerForTest`. Clearly it would be possible to do somehow, but I'm mostly thinking this would be better handled by a (future?) BGP control plane integration test.

Reply:
We do have facilities to do tests with the actual peering under [bgpv1/test](https://github.com/cilium/cilium/tree/main/pkg/bgpv1/test). What was missing there?

Reply:
@dgl Sorry, I forgot to mention you. Please elaborate on the above so that we can move this forward.

Reply:
@YutaroHayakawa I've now made a test that uses bgpv1/test, I had to make diffstore available to bgpv1/agent as well as manager to be able to get the secrets available, hopefully that's okay.